programs/bpf_loader/src/syscalls/cpi.rs

```diff
@@ -1128,29 +1359,39 @@ fn update_caller_account(
-        // with a *fixed* length, upon returning from CPI we must ensure that the
-        // current capacity is at least the original length (what is mapped in
-        // memory), so that the account's memory region never points to an
-        // invalid address.
-        let min_capacity = caller_account.original_data_len;
-        if callee_account.capacity() < min_capacity {
-            callee_account.reserve(min_capacity.saturating_sub(callee_account.capacity()))?
-        }
+    let mut zero_all_spare_capacity = false;
+    if direct_mapping {
+        if let Some(region) = account_data_region(
+            memory_mapping,
+            caller_account.vm_data_addr,
+            caller_account.original_data_len,
+        )? {
+            // Since each instruction account is directly mapped in a memory region
+            // with a *fixed* length, upon returning from CPI we must ensure that the
+            // current capacity is at least the original length (what is mapped in
+            // memory), so that the account's memory region never points to an
+            // invalid address.
+            let min_capacity = caller_account.original_data_len;
+            if callee_account.capacity() < min_capacity {
+                callee_account.reserve(min_capacity.saturating_sub(callee_account.capacity()))?;
+                zero_all_spare_capacity = true;
+            }
 
-        // If an account's data pointer has changed - because of CoW or because
-        // of using AccountSharedData directly (deprecated) - we must update the
-        // corresponding MemoryRegion in the caller's address space. Address
-        // spaces are fixed so we don't need to update the MemoryRegion's length.
-        let region = memory_mapping.region(AccessType::Load, caller_account.vm_data_addr)?;
-        let callee_ptr = callee_account.get_data().as_ptr() as u64;
-        if region.host_addr.get() != callee_ptr {
-            region.host_addr.set(callee_ptr);
+            // If an account's data pointer has changed - because of CoW, reserve() as called above
+            // or because of using AccountSharedData directly (deprecated) - we must update the
+            // corresponding MemoryRegion in the caller's address space. Address spaces are fixed so
+            // we don't need to update the MemoryRegion's length.
+            let callee_ptr = callee_account.get_data().as_ptr() as u64;
+            if region.host_addr.get() != callee_ptr {
+                region.host_addr.set(callee_ptr);
+                zero_all_spare_capacity = true;
+            }
         }
     }
-    let prev_len = *caller_account.ref_to_len_in_vm as usize;
+
+    let prev_len = *caller_account.ref_to_len_in_vm.get()? as usize;
     let post_len = callee_account.get_data().len();
+    let realloc_bytes_used = post_len.saturating_sub(caller_account.original_data_len);
     if prev_len != post_len {
```

Comment:
maybe, we should always zero spare capacity always here if `zero_all_spare_capacity == true`, regardless `post_len < prev_len`or not?

i mean, is it safe to skip zeroing conditionally while indicated to do so as such by the  flag?

say, account data vector could be changed by native program inside the called ix in the manner of Arc's Clone-on-Write fashion via `set_data_from_slice''s early-returning `.set_data()`, while still holding `prev_len == post_len`, right?

Reply:
then, assume `{prev,post}_len < original_data_len`, on-chain program can see uninitialized memory in spare capacity?

hmm, wait, to create the precondition, CoW must have been performed already before the cpi?? so, actually this is impossible?

Reply:
> i mean, is it safe to skip zeroing conditionally while indicated to do so as such by the flag?

It is safe because `slice::to_vec()` is `Vec::with_capacity(slice.len())`. So if `slice.len() < original_data_len`, it'll hit the `if callee_account.capacity() < min_capacity {` case and set `zero_all_spare_capacity=true`.

I guess I can make this more explicit and inline `slice::to_vec()` in our code.